Handle 429 Too Many Requests from Rekor

**Description**

Apparently, the team behind google-http-java-client does not want to handle 429 by default: https://github.com/googleapis/google-http-java-client/issues?q=429


Stacktrace:

```
Caused by: com.google.api.client.http.HttpResponseException: 429 Too Many Requests
    POST https://rekor.sigstore.dev/api/v1/log/entries
    <html>
    <head><title>429 Too Many Requests</title></head>
    <body>
    <center><h1>429 Too Many Requests</h1></center>
    <hr><center>nginx</center>
    </body>
    </html>
    
    	at com.google.api.client.http.HttpResponseException$Builder.build(HttpResponseException.java:293)
    	at com.google.api.client.http.HttpRequest.execute(HttpRequest.java:1118)
    	at dev.sigstore.rekor.client.RekorClient.putEntry(RekorClient.java:88)
    	at dev.sigstore.KeylessSigner.sign(KeylessSigner.java:274)
    	at dev.sigstore.KeylessSigner.signFiles(KeylessSigner.java:365)
    	at dev.sigstore.KeylessSigner.signFile(KeylessSigner.java:384)
    	at dev.sigstore.sign.work.SignWorkAction.execute(SignWorkAction.kt:57)
    	at org.gradle.workers.internal.DefaultWorkerServer.execute(DefaultWorkerServer.java:63)
    	at org.gradle.workers.internal.AbstractClassLoaderWorker$1.create(AbstractClassLoaderWorker.java:49)
    	at org.gradle.workers.internal.AbstractClassLoaderWorker$1.create(AbstractClassLoaderWorker.java:43)
    	at org.gradle.internal.classloader.ClassLoaderUtils.executeInClassloader(ClassLoaderUtils.java:97)
    	at org.gradle.workers.internal.AbstractClassLoaderWorker.executeInClassLoader(AbstractClassLoaderWorker.java:43)
    	at org.gradle.workers.internal.IsolatedClassloaderWorker.run(IsolatedClassloaderWorker.java:49)
    	at org.gradle.workers.internal.IsolatedClassloaderWorker.run(IsolatedClassloaderWorker.java:30)
    	at org.gradle.workers.internal.WorkerDaemonServer.run(WorkerDaemonServer.java:85)
    	at org.gradle.workers.internal.WorkerDaemonServer.run(WorkerDaemonServer.java:55)
    	at org.gradle.process.internal.worker.request.WorkerAction$1.call(WorkerAction.java:138)
    	at org.gradle.process.internal.worker.child.WorkerLogEventListener.withWorkerLoggingProtocol(WorkerLogEventListener.java:41)
    	at org.gradle.process.internal.worker.request.WorkerAction.run(WorkerAction.java:135)
    	at jdk.internal.reflect.GeneratedMethodAccessor45.invoke(Unknown Source)
    	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
    	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
    	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:182)
    	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:164)
    	at org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:414)
    	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
    	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
    	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:61)
```

--- a/sigstore-java/build.gradle.kts
+++ b/sigstore-java/build.gradle.kts
@@ -48,7 +48,10 @@ dependencies {
     testImplementation(platform("org.junit:junit-bom:5.9.2"))
     testImplementation("org.junit.jupiter:junit-jupiter")
 
-    testImplementation("org.mockito:mockito-core:5.1.1")
+    testImplementation(platform("org.mockito:mockito-bom:5.1.1"))
+    testImplementation("org.mockito:mockito-core")
+    testImplementation("org.mockito:mockito-junit-jupiter")
+
     testImplementation("no.nav.security:mock-oauth2-server:0.5.7")
     testImplementation("com.squareup.okhttp3:mockwebserver:4.10.0")
     testImplementation("net.sourceforge.htmlunit:htmlunit:2.70.0")

--- a/sigstore-java/src/main/java/dev/sigstore/http/HttpClients.java
+++ b/sigstore-java/src/main/java/dev/sigstore/http/HttpClients.java
@@ -15,11 +15,9 @@
  */
 package dev.sigstore.http;
 
-import com.google.api.client.http.HttpBackOffUnsuccessfulResponseHandler;
 import com.google.api.client.http.HttpRequestFactory;
 import com.google.api.client.http.HttpTransport;
 import com.google.api.client.http.apache.v2.ApacheHttpTransport;
-import com.google.api.client.util.ExponentialBackOff;
 import java.io.IOException;
 import org.apache.http.conn.ssl.NoopHostnameVerifier;
 import org.apache.http.impl.client.HttpClientBuilder;
@@ -48,7 +46,7 @@ public class HttpClients {
               request.setConnectTimeout(httpParams.getTimeout() * 1000);
               request.setReadTimeout(httpParams.getTimeout() * 1000);
               request.setUnsuccessfulResponseHandler(
-                  new HttpBackOffUnsuccessfulResponseHandler(new ExponentialBackOff()));
+                  UnsuccessfulResponseHandler.newUnsuccessfulResponseHandler());
             });
   }
 }

--- a/sigstore-java/src/main/java/dev/sigstore/http/UnsuccessfulResponseHandler.java
+++ b/sigstore-java/src/main/java/dev/sigstore/http/UnsuccessfulResponseHandler.java
@@ -1,0 +1,124 @@
+/*
+ * Copyright 2023 The Sigstore Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package dev.sigstore.http;
+
+import com.google.api.client.http.HttpRequest;
+import com.google.api.client.http.HttpResponse;
+import com.google.api.client.http.HttpUnsuccessfulResponseHandler;
+import com.google.api.client.util.BackOff;
+import com.google.api.client.util.BackOffUtils;
+import com.google.api.client.util.ExponentialBackOff;
+import com.google.api.client.util.Sleeper;
+import com.google.common.annotations.VisibleForTesting;
+import java.io.IOException;
+import java.time.Instant;
+import java.util.logging.Logger;
+import org.apache.http.client.utils.DateUtils;
+
+/**
+ * Handler for 429 and standard server errors 5XX. Objects of this class are single use. A new
+ * instance should be created for each request (but this is handled in {@link
+ * HttpClients#newRequestFactory(HttpParams)}).
+ *
+ * <p>This implementation is a bit weird, but hopefully we can just move to grpc for rekor when
+ * ready.
+ */
+public class UnsuccessfulResponseHandler implements HttpUnsuccessfulResponseHandler {
+  // handler may return before 10 retries if exponentialBackoff hits its limit first
+  private static final int MAX_RETRIES = 10;
+  private static final Logger log = Logger.getLogger(UnsuccessfulResponseHandler.class.getName());
+
+  private final Sleeper sleeper;
+  private final BackOff exponentialBackOff;
+
+  private int currentRetries = 0;
+
+  public static UnsuccessfulResponseHandler newUnsuccessfulResponseHandler() {
+    return new UnsuccessfulResponseHandler(Sleeper.DEFAULT, new ExponentialBackOff());
+  }
+
+  @VisibleForTesting
+  UnsuccessfulResponseHandler(Sleeper sleeper, ExponentialBackOff exponentialBackOff) {
+    this.sleeper = sleeper;
+    this.exponentialBackOff = exponentialBackOff;
+  }
+
+  @Override
+  public boolean handleResponse(HttpRequest request, HttpResponse response, boolean supportsRetry)
+      throws IOException {
+    if (!supportsRetry) {
+      return false;
+    }
+    if (currentRetries >= MAX_RETRIES) {
+      return false;
+    }
+    currentRetries++;
+
+    var statusCode = response.getStatusCode();
+    // we only retry 5XX and 429
+    if (!(statusCode / 100 == 5 || statusCode == 429)) {
+      return false;
+    }
+    try {
+      var headers = response.getHeaders();
+      if (headers != null) {
+        var retryAfter = headers.getRetryAfter();
+        // we only use Retry-After on 503 and 429
+        if (retryAfter != null && (statusCode == 429 || statusCode == 503)) {
+          try {
+            return handleRetryAfter(retryAfter);
+          } catch (NumberFormatException ignored) {
+            // ignored this is a parse error from Retry-After, just go to exponential backoff
+            // but maybe pop off a warning so we can inform the infrastructure owners
+            log.warning(
+                "Retry-After header in request to "
+                    + request.getUrl()
+                    + " was invalid ("
+                    + retryAfter
+                    + ")");
+          }
+        }
+      }
+      return BackOffUtils.next(sleeper, exponentialBackOff);
+    } catch (InterruptedException exception) {
+      // Mark thread as interrupted since we cannot throw InterruptedException here.
+      Thread.currentThread().interrupt();
+    }
+    return false;
+  }
+
+  // Retry-After: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Retry-After
+  private boolean handleRetryAfter(String retryAfter) throws InterruptedException, IOException {
+    var delay = calculateBackoff(retryAfter, Instant.now());
+    if (delay > 0) {
+      sleeper.sleep(delay);
+    }
+    return true;
+  }
+
+  // backoff in milliseconds
+  @VisibleForTesting
+  long calculateBackoff(String retryAfter, Instant now) throws NumberFormatException {
+    // parse as httpdate
+    var date = DateUtils.parseDate(retryAfter);
+    if (date != null) {
+      return date.toInstant().toEpochMilli() - now.toEpochMilli();
+    }
+
+    // parse as seconds (or throw NumberFormatException)
+    return Long.parseLong(retryAfter) * 1000;
+  }
+}

--- a/sigstore-java/src/test/java/dev/sigstore/http/UnsuccessfulResponseHandlerTest.java
+++ b/sigstore-java/src/test/java/dev/sigstore/http/UnsuccessfulResponseHandlerTest.java
@@ -1,0 +1,156 @@
+/*
+ * Copyright 2023 The Sigstore Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package dev.sigstore.http;
+
+import com.google.api.client.http.HttpHeaders;
+import com.google.api.client.http.HttpResponse;
+import com.google.api.client.util.ExponentialBackOff;
+import com.google.api.client.util.Sleeper;
+import java.io.IOException;
+import java.time.Instant;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.ValueSource;
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.Mockito;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+@ExtendWith(MockitoExtension.class)
+class UnsuccessfulResponseHandlerTest {
+  private static final Instant testNow =
+      Instant.ofEpochMilli(1675349494000L); // February 2, 2023 2:51:34 PM GMT
+  @Mock private Sleeper sleeper;
+  @Mock private ExponentialBackOff exponentialBackOff;
+  @InjectMocks private UnsuccessfulResponseHandler testHandler;
+
+  @Mock private HttpResponse httpResponse;
+  @Mock private HttpHeaders httpHeaders;
+
+  @Test
+  public void testHandleResponse_retryLimit() throws IOException, InterruptedException {
+    int count = 0;
+    Mockito.when(exponentialBackOff.nextBackOffMillis()).thenReturn(10L);
+    Mockito.when(httpResponse.getStatusCode()).thenReturn(500);
+    while (testHandler.handleResponse(null, httpResponse, true)) {
+      count++;
+    }
+
+    Assertions.assertEquals(10, count);
+  }
+
+  @ParameterizedTest
+  @ValueSource(ints = {500, 509})
+  public void testHandleResponse_5xx(int status) throws IOException, InterruptedException {
+    Mockito.when(exponentialBackOff.nextBackOffMillis()).thenReturn(10L);
+    Mockito.when(httpResponse.getStatusCode()).thenReturn(status);
+
+    var resp = testHandler.handleResponse(null, httpResponse, true);
+    Assertions.assertTrue(resp);
+
+    Mockito.verify(exponentialBackOff).nextBackOffMillis();
+    Mockito.verify(sleeper).sleep(10L);
+    Mockito.verifyNoMoreInteractions(sleeper, exponentialBackOff);
+  }
+
+  @ParameterizedTest
+  @ValueSource(ints = {500, 509})
+  public void testHandleResponse_5xxIgnoreRetryAfter(int status)
+      throws IOException, InterruptedException {
+    Mockito.when(exponentialBackOff.nextBackOffMillis()).thenReturn(10L);
+    Mockito.when(httpResponse.getStatusCode()).thenReturn(status);
+    Mockito.when(httpResponse.getHeaders()).thenReturn(httpHeaders);
+    Mockito.when(httpHeaders.getRetryAfter()).thenReturn("50");
+
+    var resp = testHandler.handleResponse(null, httpResponse, true);
+    Assertions.assertTrue(resp);
+
+    Mockito.verify(exponentialBackOff).nextBackOffMillis();
+    Mockito.verify(sleeper).sleep(10L);
+    Mockito.verifyNoMoreInteractions(sleeper, exponentialBackOff);
+  }
+
+  @ParameterizedTest
+  @ValueSource(ints = {429, 503})
+  public void testHandleResponse_withRetryAfter(int status)
+      throws IOException, InterruptedException {
+    Mockito.when(httpResponse.getStatusCode()).thenReturn(status);
+    Mockito.when(httpResponse.getHeaders()).thenReturn(httpHeaders);
+    Mockito.when(httpHeaders.getRetryAfter()).thenReturn("50");
+
+    var resp = testHandler.handleResponse(null, httpResponse, true);
+    Assertions.assertTrue(resp);
+
+    Mockito.verify(sleeper).sleep(50000L);
+    Mockito.verifyNoMoreInteractions(sleeper, exponentialBackOff);
+  }
+
+  @ParameterizedTest
+  @ValueSource(ints = {429, 503})
+  public void testHandleResponse_withoutRetryAfter(int status)
+      throws IOException, InterruptedException {
+    Mockito.when(httpResponse.getStatusCode()).thenReturn(status);
+    Mockito.when(httpResponse.getHeaders()).thenReturn(httpHeaders);
+    Mockito.when(exponentialBackOff.nextBackOffMillis()).thenReturn(10L);
+
+    var resp = testHandler.handleResponse(null, httpResponse, true);
+    Assertions.assertTrue(resp);
+
+    Mockito.verify(exponentialBackOff).nextBackOffMillis();
+    Mockito.verify(sleeper).sleep(10L);
+    Mockito.verifyNoMoreInteractions(sleeper, exponentialBackOff);
+  }
+
+  @Test
+  public void testHandleResponse_noRetry() throws IOException {
+    var resp = testHandler.handleResponse(null, null, false);
+    Assertions.assertFalse(resp);
+    Mockito.verifyNoInteractions(sleeper, exponentialBackOff);
+  }
+
+  @Test
+  public void testHandleResponse_noRetryCode() throws IOException {
+    Mockito.when(httpResponse.getStatusCode()).thenReturn(404);
+
+    var resp = testHandler.handleResponse(null, httpResponse, true);
+    Assertions.assertFalse(resp);
+
+    Mockito.verify(httpResponse, Mockito.times(1)).getStatusCode();
+    Mockito.verifyNoInteractions(sleeper, exponentialBackOff);
+  }
+
+  @Test
+  public void testCalculateBackoff_httpDate() {
+    String httpDate = "Thu, 2 Feb 2023 14:51:54 GMT"; // 20 seconds after "testNow"
+    var result = testHandler.calculateBackoff(httpDate, testNow);
+    Assertions.assertEquals(20 * 1000, result);
+  }
+
+  @Test
+  public void testCalculateBackoff_seconds() {
+    String seconds = "42";
+    var result = testHandler.calculateBackoff(seconds, /* ignored */ testNow);
+    Assertions.assertEquals(42 * 1000, result);
+  }
+
+  @Test
+  public void testCalculateBackoff_badInput() {
+    Assertions.assertThrows(
+        NumberFormatException.class, () -> testHandler.calculateBackoff("blah", testNow));
+  }
+}

